File: sdk-node/src/eventBuilder.ts
import { RHEONICValidationError } from "./providerModelValidation.js";

export interface EventRequest {
  endpoint?: string;
  feature?: string;
  request_fingerprint?: string;
  input_tokens?: number;
  input_tokens_estimate?: number;
  token_explosion_tokens?: number;
  max_output_tokens?: number;
  protect_decision?: string;
  protect_reason?: string;
}

export interface EventResponse {
  http_status?: number;
  latency_ms?: number;
  output_tokens?: number;
  total_tokens?: number;
  error_type?: string;
  error_message?: string;
}

export interface EventPayload {
  ts: string;
  provider: string;
  requested_model: string | null;
  resolved_model: string | null;
  environment: string;
  status?: string;
  request: EventRequest;
  response: EventResponse;
}

export interface BuildEventInput {
  provider: string;
  model?: string | null;
  environment?: string;
  ts?: string;
  status?: string;
  request?: EventRequest;
  response?: EventResponse;
}

export interface InternalEventInput {
  provider: string;
  requested_model?: string | null;
  resolved_model?: string | null;
  environment?: string;
  ts?: string;
  status?: string;
  request?: EventRequest;
  response?: EventResponse;
}

const PUBLIC_EVENT_KEYS = new Set(["provider", "model", "environment", "ts", "status", "request", "response"]);
const INTERNAL_EVENT_KEYS = new Set([
  "provider",
  "requested_model",
  "resolved_model",
  "environment",
  "ts",
  "status",
  "request",
  "response",
]);
const REQUEST_KEYS = new Set([
  "endpoint",
  "feature",
  "request_fingerprint",
  "input_tokens",
  "input_tokens_estimate",
  "token_explosion_tokens",
  "max_output_tokens",
  "protect_decision",
  "protect_reason",
]);
const RESPONSE_KEYS = new Set(["http_status", "latency_ms", "output_tokens", "total_tokens", "error_type", "error_message"]);

function failValidation(message: string): never {
  throw new RHEONICValidationError(message, "event", "", []);
}

function assertPlainObject(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    failValidation(`RHEONIC: ${label} must be an object.`);
  }
  return value as Record<string, unknown>;
}

function assertNoExtraKeys(record: Record<string, unknown>, allowedKeys: Set<string>, label: string): void {
  for (const key of Object.keys(record)) {
    if (!allowedKeys.has(key)) {
      failValidation(`RHEONIC: unexpected ${label} property: ${key}`);
    }
  }
}

function assertOptionalString(value: unknown, label: string): void {
  if (value !== undefined && value !== null && typeof value !== "string") {
    failValidation(`RHEONIC: ${label} must be a string.`);
  }
}

function assertOptionalNumber(value: unknown, label: string): void {
  if (value !== undefined && value !== null && typeof value !== "number") {
    failValidation(`RHEONIC: ${label} must be a number.`);
  }
}

function validateRequestShape(request: unknown): EventRequest {
  const record = assertPlainObject(request, "event.request");
  assertNoExtraKeys(record, REQUEST_KEYS, "event.request");
  assertOptionalString(record.endpoint, "event.request.endpoint");
  assertOptionalString(record.feature, "event.request.feature");
  assertOptionalString(record.request_fingerprint, "event.request.request_fingerprint");
  assertOptionalNumber(record.input_tokens, "event.request.input_tokens");
  assertOptionalNumber(record.input_tokens_estimate, "event.request.input_tokens_estimate");
  assertOptionalNumber(record.token_explosion_tokens, "event.request.token_explosion_tokens");
  assertOptionalNumber(record.max_output_tokens, "event.request.max_output_tokens");
  assertOptionalString(record.protect_decision, "event.request.protect_decision");
  assertOptionalString(record.protect_reason, "event.request.protect_reason");
  return record as EventRequest;
}

function validateResponseShape(response: unknown): EventResponse {
  const record = assertPlainObject(response, "event.response");
  assertNoExtraKeys(record, RESPONSE_KEYS, "event.response");
  assertOptionalNumber(record.http_status, "event.response.http_status");
  assertOptionalNumber(record.latency_ms, "event.response.latency_ms");
  assertOptionalNumber(record.output_tokens, "event.response.output_tokens");
  assertOptionalNumber(record.total_tokens, "event.response.total_tokens");
  assertOptionalString(record.error_type, "event.response.error_type");
  assertOptionalString(record.error_message, "event.response.error_message");
  return record as EventResponse;
}

function validateBuildEventInput(input: BuildEventInput): BuildEventInput {
  const record = assertPlainObject(input, "event");
  assertNoExtraKeys(record, PUBLIC_EVENT_KEYS, "event");
  if (typeof record.provider !== "string" || record.provider.trim().length === 0) {
    failValidation("RHEONIC: event.provider must be a non-empty string.");
  }
  assertOptionalString(record.model, "event.model");
  assertOptionalString(record.environment, "event.environment");
  assertOptionalString(record.ts, "event.ts");
  assertOptionalString(record.status, "event.status");
  if (record.request !== undefined) {
    validateRequestShape(record.request);
  }
  if (record.response !== undefined) {
    validateResponseShape(record.response);
  }
  return input;
}

function validateInternalEventInput(input: InternalEventInput): InternalEventInput {
  const record = assertPlainObject(input, "event");
  assertNoExtraKeys(record, INTERNAL_EVENT_KEYS, "event");
  if (typeof record.provider !== "string" || record.provider.trim().length === 0) {
    failValidation("RHEONIC: event.provider must be a non-empty string.");
  }
  assertOptionalString(record.requested_model, "event.requested_model");
  assertOptionalString(record.resolved_model, "event.resolved_model");
  assertOptionalString(record.environment, "event.environment");
  assertOptionalString(record.ts, "event.ts");
  assertOptionalString(record.status, "event.status");
  if (record.request !== undefined) {
    validateRequestShape(record.request);
  }
  if (record.response !== undefined) {
    validateResponseShape(record.response);
  }
  return input;
}

export function validateEventPayload(input: EventPayload): EventPayload {
  const record = assertPlainObject(input, "event");
  assertNoExtraKeys(record, INTERNAL_EVENT_KEYS, "event");
  for (const requiredKey of ["ts", "provider", "requested_model", "resolved_model", "environment", "request", "response"]) {
    if (!Object.prototype.hasOwnProperty.call(record, requiredKey)) {
      failValidation(`RHEONIC: event.${requiredKey} is required.`);
    }
  }
  if (typeof record.ts !== "string" || record.ts.trim().length === 0) {
    failValidation("RHEONIC: event.ts must be a non-empty string.");
  }
  if (typeof record.provider !== "string" || record.provider.trim().length === 0) {
    failValidation("RHEONIC: event.provider must be a non-empty string.");
  }
  if (typeof record.environment !== "string" || record.environment.trim().length === 0) {
    failValidation("RHEONIC: event.environment must be a non-empty string.");
  }
  assertOptionalString(record.requested_model, "event.requested_model");
  assertOptionalString(record.resolved_model, "event.resolved_model");
  assertOptionalString(record.status, "event.status");
  validateRequestShape(record.request);
  validateResponseShape(record.response);
  return input;
}

export function normalizeEventPayload(input: EventPayload | BuildEventInput): EventPayload {
  const record = assertPlainObject(input, "event");
  const isInternal =
    Object.prototype.hasOwnProperty.call(record, "requested_model")
    || Object.prototype.hasOwnProperty.call(record, "resolved_model");
  if (isInternal) {
    return validateEventPayload(input as EventPayload);
  }
  return buildEvent(input as BuildEventInput);
}

export function buildInternalEvent(input: InternalEventInput): EventPayload {
  validateInternalEventInput(input);
  return {
    ts: input.ts ?? new Date().toISOString(),
    provider: input.provider,
    requested_model: input.requested_model ?? null,
    resolved_model: input.resolved_model ?? null,
    environment: input.environment ?? "dev",
    ...(input.status !== undefined ? { status: input.status } : {}),
    request: input.request ?? {},
    response: input.response ?? {},
  };
}

export function buildEvent(input: BuildEventInput): EventPayload {
  validateBuildEventInput(input);
  return buildInternalEvent({
    provider: input.provider,
    requested_model: input.model ?? null,
    resolved_model: null,
    environment: input.environment,
    ts: input.ts,
    status: input.status,
    request: input.request,
    response: input.response,
  });
}

export class EventBuilder {
  public build(payload: BuildEventInput): EventPayload {
    return buildEvent(payload);
  }
}

File: sdk-python/src/rheonic/event_builder.py
# Event builder helpers.
from datetime import datetime, timezone
from typing import Any, cast

from rheonic.logger import get_logger
from rheonic.protect_engine import RHEONICValidationError

logger = get_logger(__name__)

_PUBLIC_EVENT_KEYS = {"provider", "model", "environment", "ts", "status", "request", "response"}
_INTERNAL_EVENT_KEYS = {
    "ts",
    "provider",
    "requested_model",
    "resolved_model",
    "environment",
    "status",
    "request",
    "response",
}
_REQUEST_KEYS = {
    "endpoint",
    "feature",
    "request_fingerprint",
    "input_tokens",
    "input_tokens_estimate",
    "token_explosion_tokens",
    "max_output_tokens",
    "protect_decision",
    "protect_reason",
}
_RESPONSE_KEYS = {"http_status", "latency_ms", "total_tokens", "output_tokens", "error_type", "error_message"}


def _fail_validation(message: str) -> None:
    raise RHEONICValidationError(message)


def _validate_mapping(value: dict[str, Any] | None, allowed_keys: set[str], label: str) -> dict[str, Any]:
    if value is None:
        return {}
    if not isinstance(value, dict):
        _fail_validation(f"RHEONIC: {label} must be an object.")
    extra_keys = set(value) - allowed_keys
    if extra_keys:
        _fail_validation(f"RHEONIC: unexpected {label} property: {sorted(extra_keys)[0]}")
    return value


def _validate_optional_string(value: Any, label: str) -> None:
    if value is not None and not isinstance(value, str):
        _fail_validation(f"RHEONIC: {label} must be a string.")


def _validate_optional_number(value: Any, label: str) -> None:
    if value is not None and not isinstance(value, int):
        _fail_validation(f"RHEONIC: {label} must be a number.")


def _validate_request(request: dict[str, Any] | None) -> dict[str, Any]:
    request_map = _validate_mapping(request, _REQUEST_KEYS, "event.request")
    _validate_optional_string(request_map.get("endpoint"), "event.request.endpoint")
    _validate_optional_string(request_map.get("feature"), "event.request.feature")
    _validate_optional_string(request_map.get("request_fingerprint"), "event.request.request_fingerprint")
    _validate_optional_number(request_map.get("input_tokens"), "event.request.input_tokens")
    _validate_optional_number(request_map.get("input_tokens_estimate"), "event.request.input_tokens_estimate")
    _validate_optional_number(request_map.get("token_explosion_tokens"), "event.request.token_explosion_tokens")
    _validate_optional_number(request_map.get("max_output_tokens"), "event.request.max_output_tokens")
    _validate_optional_string(request_map.get("protect_decision"), "event.request.protect_decision")
    _validate_optional_string(request_map.get("protect_reason"), "event.request.protect_reason")
    return request_map


def _validate_response(response: dict[str, Any] | None) -> dict[str, Any]:
    response_map = _validate_mapping(response, _RESPONSE_KEYS, "event.response")
    _validate_optional_number(response_map.get("http_status"), "event.response.http_status")
    _validate_optional_number(response_map.get("latency_ms"), "event.response.latency_ms")
    _validate_optional_number(response_map.get("total_tokens"), "event.response.total_tokens")
    _validate_optional_number(response_map.get("output_tokens"), "event.response.output_tokens")
    _validate_optional_string(response_map.get("error_type"), "event.response.error_type")
    _validate_optional_string(response_map.get("error_message"), "event.response.error_message")
    return response_map


def build_event(
    provider: str,
    model: str | None = None,
    environment: str = "dev",
    status: str | None = None,
    request: dict[str, Any] | None = None,
    response: dict[str, Any] | None = None,
    ts: str | None = None,
) -> dict[str, Any]:
    # Build backend ingest payload without project_id.
    if not isinstance(provider, str) or not provider.strip():
        _fail_validation("RHEONIC: event.provider must be a non-empty string.")
    _validate_optional_string(model, "event.model")
    _validate_optional_string(environment, "event.environment")
    _validate_optional_string(status, "event.status")
    _validate_optional_string(ts, "event.ts")
    return {
        "ts": ts or datetime.now(timezone.utc).isoformat(),
        "provider": provider,
        "requested_model": model,
        "resolved_model": None,
        "environment": environment,
        **({"status": status} if status is not None else {}),
        "request": _validate_request(request),
        "response": _validate_response(response),
    }


def build_internal_event(
    provider: str,
    requested_model: str | None = None,
    resolved_model: str | None = None,
    environment: str = "dev",
    status: str | None = None,
    request: dict[str, Any] | None = None,
    response: dict[str, Any] | None = None,
    ts: str | None = None,
) -> dict[str, Any]:
    if not isinstance(provider, str) or not provider.strip():
        _fail_validation("RHEONIC: event.provider must be a non-empty string.")
    _validate_optional_string(requested_model, "event.requested_model")
    _validate_optional_string(resolved_model, "event.resolved_model")
    _validate_optional_string(environment, "event.environment")
    _validate_optional_string(status, "event.status")
    _validate_optional_string(ts, "event.ts")
    return {
        "ts": ts or datetime.now(timezone.utc).isoformat(),
        "provider": provider,
        "requested_model": requested_model,
        "resolved_model": resolved_model,
        "environment": environment,
        **({"status": status} if status is not None else {}),
        "request": _validate_request(request),
        "response": _validate_response(response),
    }


def normalize_event_payload(payload: dict[str, Any]) -> dict[str, Any]:
    if not isinstance(payload, dict):
        _fail_validation("RHEONIC: event must be an object.")

    payload_keys = set(payload)
    if {"requested_model", "resolved_model"} & payload_keys:
        extra_keys = payload_keys - _INTERNAL_EVENT_KEYS
        if extra_keys:
            _fail_validation(f"RHEONIC: unexpected event property: {sorted(extra_keys)[0]}")
        missing_keys = [
            key
            for key in ("ts", "provider", "requested_model", "resolved_model", "environment", "request", "response")
            if key not in payload
        ]
        if missing_keys:
            _fail_validation(f"RHEONIC: event.{missing_keys[0]} is required.")
        ts_value = payload.get("ts")
        provider_value = payload.get("provider")
        requested_model = payload.get("requested_model")
        resolved_model = payload.get("resolved_model")
        environment_value = payload.get("environment")
        if not isinstance(ts_value, str) or not ts_value.strip():
            _fail_validation("RHEONIC: event.ts must be a non-empty string.")
        if not isinstance(provider_value, str) or not provider_value.strip():
            _fail_validation("RHEONIC: event.provider must be a non-empty string.")
        if not isinstance(environment_value, str) or not environment_value.strip():
            _fail_validation("RHEONIC: event.environment must be a non-empty string.")
        _validate_optional_string(payload.get("status"), "event.status")
        internal_ts = cast(str, ts_value)
        internal_provider = cast(str, provider_value)
        internal_environment = cast(str, environment_value)
        return build_internal_event(
            provider=internal_provider,
            requested_model=requested_model if isinstance(requested_model, str) else None,
            resolved_model=resolved_model if isinstance(resolved_model, str) else None,
            environment=internal_environment,
            status=payload.get("status") if isinstance(payload.get("status"), str) else None,
            request=payload.get("request") if isinstance(payload.get("request"), dict) else None,
            response=payload.get("response") if isinstance(payload.get("response"), dict) else None,
            ts=internal_ts,
        )

    extra_keys = payload_keys - _PUBLIC_EVENT_KEYS
    if extra_keys:
        _fail_validation(f"RHEONIC: unexpected event property: {sorted(extra_keys)[0]}")
    provider_value = payload.get("provider")
    environment_value = payload.get("environment", "dev")
    if not isinstance(provider_value, str) or not provider_value.strip():
        _fail_validation("RHEONIC: event.provider must be a non-empty string.")
    if not isinstance(environment_value, str) or not environment_value.strip():
        _fail_validation("RHEONIC: event.environment must be a non-empty string.")
    _validate_optional_string(payload.get("status"), "event.status")
    public_provider = cast(str, provider_value)
    public_environment = cast(str, environment_value)
    return build_event(
        provider=public_provider,
        model=payload.get("model"),
        environment=public_environment,
        status=payload.get("status") if isinstance(payload.get("status"), str) else None,
        request=payload.get("request") if isinstance(payload.get("request"), dict) else None,
        response=payload.get("response") if isinstance(payload.get("response"), dict) else None,
        ts=payload.get("ts") if isinstance(payload.get("ts"), str) else None,
    )


class EventBuilder:
    # Builds normalized usage events from provider responses.

    def build(self, payload: dict[str, Any]) -> dict[str, Any]:
        # Build a backend-compatible event payload.
        try:
            return build_event(
                provider=str(payload.get("provider", "unknown")),
                model=payload.get("model") if isinstance(payload.get("model"), str) else None,
                environment=str(payload.get("environment", "dev")),
                status=payload.get("status") if isinstance(payload.get("status"), str) else None,
                request=payload.get("request") if isinstance(payload.get("request"), dict) else None,
                response=payload.get("response") if isinstance(payload.get("response"), dict) else None,
                ts=payload.get("ts") if isinstance(payload.get("ts"), str) else None,
            )
        except Exception:
            logger.exception("Event builder failed")
            raise
